Added benchmarks project

RelaxSpirit opened the pull request:
## Description
Added the benchmark project.
Fix README samples and benchmark description.

## Motivation and Context
Visual benchmark results are needed.
Updating the README.

## How Has This Been Tested?
Benchmarks in virtual nanoDevice.

## Types of changes
<!--- What types of changes does this PR introduce? Put an `x` in all the boxes that apply: -->
- [ ] Improvement (non-breaking change that improves a feature, code or algorithm)
- [ ] Bug fix (non-breaking change which fixes an issue with code or algorithm)
- [ ] New feature (non-breaking change which adds functionality to code)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Config and build (change in the configuration and build system, has no impact on code or features)
- [ ] Dependencies (update dependencies and changes associated, has no impact on code or features)
- [ ] Unit Tests (add new Unit Test(s) or improved existing one(s), has no impact on code or features)
- [x] Documentation (changes or updates in the documentation, has no impact on code or features)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
<!--- PLEASE PLEASE PLEASE don't tick all of them just because -->
- [x] My code follows the code style of this project (only if there are changes in source code).
- [ ] My changes require an update to the documentation (there are changes that require the docs website to be updated).
- [x] I have updated the documentation accordingly (the changes require an update on the docs in this repo).
- [x] I have read the [CONTRIBUTING](https://github.com/nanoframework/.github/blob/main/CONTRIBUTING.md) document.
- [x] I have tested everything locally and all new and existing tests passed (only if there are changes in source code).
- [ ] I have added new tests to cover my changes.



<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **Chores**
  - Added a lock file to specify exact versions of dependencies for improved consistency and reproducibility in the project.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

RelaxSpirit committed:
Removed unnecessary memory allocations.
Fix deserialization of a single two-dimensional array and hashtable.
Add and updated UnitTest.

RelaxSpirit committed:
Added benchmark project.
Update README.

josesimoes committed:
Merge branch 'main' into unnecessary_memory_allocations

josesimoes committed:
dummy change

RelaxSpirit committed:
Merge branch 'main' into Benchmarks_2

RelaxSpirit committed:
Fix UnitTest

RelaxSpirit committed:
Merge branch 'unnecessary_memory_allocations' into Benchmarks_2

RelaxSpirit committed:
Сode review by PR

RelaxSpirit committed:
Fix сasting a signed num types

RelaxSpirit committed:
Merge branch 'unnecessary_memory_allocations' into Benchmarks_2

RelaxSpirit committed:
Merge branch 'main' into Benchmarks_2

RelaxSpirit committed:
Fix project references

josesimoes commented:
/AzurePipelines run

josesimoes commented on nanoFramework.MessagePack.Benchmark/DeserializationBenchmarks/ComparativeDeserializationBenchmark.cs:
```diff
@@ -0,0 +1,46 @@
+    [IterationCount(10)]
+    internal class ComparativeDeserializationBenchmark : BaseIterationBenchmark
+    {
+        protected override int IterationCount => 1;
```
The project style calls for this pattern for private fields:
```suggestion
        protected override int _iterationCount => 1;
```

If you add a ref to the StyleCop nuget (see the library project) Visual Studio takes care of this for your and lets you know when something is wrong.

josesimoes replied:
For reference, here is the code style guide:
https://docs.nanoframework.net/content/contributing/cs-coding-style.html

RelaxSpirit replied:
@josesimoes I'm sorry, all this has moved from a similar project, but for Json :)
I'll fix everything.

josesimoes replied:
No worries, nothing to apologise. You're doing a great work bringing MessagePack to the nanoFramework community.
We have a high bar on code quality. These are tiny nits that need to be addressed in order to maintain that level.

josesimoes commented on README.md:
```diff
@@ -18,6 +18,42 @@ This repository contains the MessagePack library for the .NET **nanoFramework**.
 
 MessagePack is a simple, lightweight serialization library, inspired by [MsgPack.Light](https://github.com/progaudi/MsgPack.Light), that can be used in .NET[nanoFramework](https://github.com/nanoframework) solutions.
 
+# Approximate measurement results for different serializers
```
```suggestion
## Benchmarks
```

And please move the section and place it after the "Usage" one. This is important when one is comparing serializer options, but secondary to the usage explanations.

josesimoes commented on README.md:
````diff
@@ -18,6 +18,42 @@ This repository contains the MessagePack library for the .NET **nanoFramework**.
+| MessagePackSerializationBenchmark | 10             | 13.8 ms | 12 ms | 18 ms |
+| ---------------------------------------------------------------------------- |
+```
+In terms of speed, MessagePack is no worse than a Json serializer, and slightly wins in terms of the size of the serialized data
````
```suggestion
As it can be seen from the benchmark results above, in what concerns speed and compaction, `MessagePack` performs better than the Json serializer. Comming at no surprise, Binary serialization is the most performant one.
```

josesimoes reviewed (commented):
All good, just some issues with code style and a tweak in the readme.

RelaxSpirit committed:
Fix README.
Add StyleCop.
Code review, an attempt to fulfill all the requirements of a code style.

RelaxSpirit committed:
Update NuGet

josesimoes commented:
@nfbot updatedependencies

josesimoes committed:
Merge branch 'main' into Benchmarks_2

josesimoes committed:
Fix code style in benchmark project

RelaxSpirit committed:
Merge branch 'nanoframework:main' into Benchmarks_2

josesimoes committed:
Update refe to json

josesimoes committed:
Fix usings inclusion

josesimoes committed:
Merge branch 'main' of https://github.com/nanoframework/nanoFramework.MessagePack into Benchmarks_2

josesimoes reviewed (approved):
LGTM! 
I've fixed some left nits.
Thanks for adding the benchmark project. 
Now we have a completely documented alternative library!

josesimoes commented:
/AzurePipelines run

RelaxSpirit commented:
> LGTM! I've fixed some left nits. Thanks for adding the benchmark project. Now we have a completely documented alternative library!

Thank you for your help! The most difficult thing is to put the code in order, not any creativity :))
Of course, there are still points in the library that need improvement, but this is already an order of necessity :)
